src/Symfony/Component/Security/Http/Authenticator/Passport/Badge/UserBadge.php

```diff
@@ -0,0 +1,63 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Security\Http\Authenticator\Passport\Badge;
+
+use Symfony\Component\Security\Core\Exception\UsernameNotFoundException;
+use Symfony\Component\Security\Core\User\UserInterface;
+use Symfony\Component\Security\Http\EventListener\UserProviderListener;
+
+class UserBadge implements BadgeInterface
+{
+    private $username;
+    private $userLoader;
+    private $user;
+
+    public function __construct(?string $username, ?callable $userLoader = null)
```

Comment:
We should probably document both of these, because:

1) We need to communicate that `$username` is not necessarily a username... but just some "identifier". We probably need to communicate its *purpose*. For example, suppose I have a login form that has an `email` field AND a `company ` drop-down. So, when I query the database for the user, the query is something like:

> SELECT * FROM user WHERE email = :email AND company_id = :companyId

Obviously, I would probably make the query and load the user by passing a 2nd argument to `new UserBadge()` with a callback that makes the custom query.

The question is: what value should I pass as the `$username`? Should it just be `$email`? Probably not. Should it be `['email' => $email, 'companyId' => $companyId]`? Well, right now, the first argument must be a string... and probably (?) we want to keep it that way. So, should the user pass `$email'._'.$companyId`... because the purpose of this argument is to be some "unique string identifier" of the user?

2) We should document that the 2nd arg is optional, must return `UserInterface` or `null` and will be passed the `$username`.